docs/configuration.md
# Configuration

The configuration is saved in the database, you can change and view them via the admin interface.

You can find them in the appconfig table.

```sql
mysql> select * from oc_appconfig where appid = 'news';
+-------+--------------------------+-------------+
| appid | configkey                | configvalue |
+-------+--------------------------+-------------+
| news  | autoPurgeCount           | 200         |
| news  | autoPurgeMinimumInterval | 60          |
| news  | enabled                  | yes         |
| news  | exploreUrl               |             |
| news  | feedFetcherTimeout       | 60          |
| news  | installed_version        | 15.0.0      |
| news  | maxRedirects             | 10          |
| news  | types                    |             |
| news  | updateInterval           | 3600        |
| news  | useCronUpdates           | 1           |
+-------+--------------------------+-------------+
```

# Legacy Configuration before News 15.0.0

All configuration values are set inside **nextcloud/data/news/config/config.ini** and can be edited in the admin panel.

The configuration is in **INI** format and looks like this:

```ini
autoPurgeMinimumInterval = 60
autoPurgeCount = 200
maxRedirects = 10
maxSize = 104857600
feedFetcherTimeout = 60
useCronUpdates = true
exploreUrl =
```


* **autoPurgeMinimumInterval**: Minimum amount of seconds after deleted feeds and folders are removed from the database. Values below 60 seconds are ignored
* **autoPurgeCount**: Defines the maximum amount of articles that can be read per feed which won't be deleted by the cleanup job; if old articles reappear after being read, increase this value; negative values such as -1 will turn this feature off completely
* **maxRedirects**: How many redirects the updater should follow
* **maxSize**: Maximum feed size in bytes. If the RSS/Atom page is bigger than this value, the update will be aborted
* **feedFetcherTimeout**: Maximum number of seconds to wait for an RSS or Atom feed to load. If a feed takes longer than that number of seconds to update, the update will be aborted
* **useCronUpdates**: To use a custom update/cron script you need to disable the cronjob which is run by Nextcloud by default by setting this to false
* **exploreUrl**: If given that url will be contacted for fetching content for the explore feed

news
/srv/http/apps/news/

lib/Config/Config.php
<?php
/**
 * Nextcloud - News
 *
 * This file is licensed under the Affero General Public License version 3 or
 * later. See the COPYING file.
 *
 * @author    Alessandro Cosentino <cosenal@gmail.com>
 * @author    Bernhard Posselt <dev@bernhard-posselt.com>
 * @copyright 2012 Alessandro Cosentino
 * @copyright 2012-2014 Bernhard Posselt
 */

namespace OCA\News\Config;

use OCA\News\Utility\PsrLogger;
use OCP\Files\Folder;

class Config
{

    private $fileSystem;
    private $autoPurgeMinimumInterval;  // seconds, used to define how
                                        // long deleted folders and feeds
                                        // should still be kept for an
                                        // undo actions
    private $autoPurgeCount;  // number of allowed unread articles per feed
    private $maxRedirects;  // seconds
    private $feedFetcherTimeout;  // seconds
    private $useCronUpdates;  // turn off updates run by the cron
    private $logger;
    private $loggerParams;
    private $maxSize;
    private $exploreUrl;
    private $updateInterval;

    public function __construct(
        Folder $fileSystem,
        PsrLogger $logger,
        $LoggerParameters
    ) {
        $this->fileSystem = $fileSystem;
        $this->autoPurgeMinimumInterval = 60;
        $this->autoPurgeCount = 200;
        $this->maxRedirects = 10;
        $this->maxSize = 100 * 1024 * 1024; // 100Mb
        $this->feedFetcherTimeout = 60;
        $this->useCronUpdates = true;
        $this->logger = $logger;
        $this->exploreUrl = '';
        $this->loggerParams = $LoggerParameters;
        $this->updateInterval = 3600;
    }

    public function getAutoPurgeMinimumInterval()
    {
        if ($this->autoPurgeMinimumInterval > 60) {
            return $this->autoPurgeMinimumInterval;
        } else {
            return 60;
        }
    }

    public function getAutoPurgeCount()
    {
        return $this->autoPurgeCount;
    }


    public function getMaxRedirects()
    {
        return $this->maxRedirects;
    }


    public function getFeedFetcherTimeout()
    {
        return $this->feedFetcherTimeout;
    }


    public function getUseCronUpdates()
    {
        return $this->useCronUpdates;
    }


    public function getMaxSize()
    {
        return $this->maxSize;
    }


    public function getExploreUrl()
    {
        return $this->exploreUrl;
    }

    public function getUpdateInterval()
    {
        return $this->updateInterval;
    }

    public function setAutoPurgeMinimumInterval($value)
    {
        $this->autoPurgeMinimumInterval = $value;
    }


    public function setAutoPurgeCount($value)
    {
        $this->autoPurgeCount = $value;
    }


    public function setMaxRedirects($value)
    {
        $this->maxRedirects = $value;
    }


    public function setFeedFetcherTimeout($value)
    {
        $this->feedFetcherTimeout = $value;
    }


    public function setUseCronUpdates($value)
    {
        $this->useCronUpdates = $value;
    }

    public function setMaxSize($value)
    {
        $this->maxSize = $value;
    }


    public function setExploreUrl($value)
    {
        $this->exploreUrl = $value;
    }

    public function setUpdateInterval($value)
    {
        $this->updateInterval = $value;
    }



    public function read($configPath, $createIfNotExists = false)
    {
        if ($createIfNotExists && !$this->fileSystem->nodeExists($configPath)) {
            $this->fileSystem->newFile($configPath);
            $this->write($configPath);
        } else {
            $content = $this->fileSystem->get($configPath)->getContent();
            $configValues = parse_ini_string($content);

            if ($configValues === false || count($configValues) === 0) {
                $this->logger->warning(
                    'Configuration invalid. Ignoring values.',
                    $this->loggerParams
                );
            } else {
                foreach ($configValues as $key => $value) {
                    if (property_exists($this, $key)) {
                        $type = gettype($this->$key);
                        settype($value, $type);
                        $this->$key = $value;
                    } else {
                        $this->logger->warning(
                            'Configuration value "' . $key .
                            '" does not exist. Ignored value.',
                            $this->loggerParams
                        );
                    }
                }
            }
        }
    }


    public function write($configPath)
    {
        $ini =
            'autoPurgeMinimumInterval = ' .
                $this->autoPurgeMinimumInterval . "\n" .
            'autoPurgeCount = ' .
                $this->autoPurgeCount . "\n" .
            'maxRedirects = ' .
                $this->maxRedirects . "\n" .
            'maxSize = ' .
                $this->maxSize . "\n" .
            'exploreUrl = ' .
                $this->exploreUrl . "\n" .
            'feedFetcherTimeout = ' .
                $this->feedFetcherTimeout . "\n" .
            'updateInterval = ' .
                $this->updateInterval . "\n" .
            'useCronUpdates = ' .
                var_export($this->useCronUpdates, true);
        ;

        $this->fileSystem->get($configPath)->putContent($ini);
    }
}
